tsdb/shard.go

```diff
@@ -1786,154 +1822,255 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
 		} else {
-			return nil, nil
+			return nil
 		}
+	})()
+
+	// Read all the pending field and measurement write or delete
+	// requests
+	for {
+		select {
+		case wr := <-fscm.writeRequests:
+			changes = append(changes, wr.changes)
+			errorChannels = append(errorChannels, wr.errorReturn)
+			continue
+		default:
+		}
+		break
+	}
+	// marshal the alice of slices of field changes
+	var b []byte
+	b, err = marshalFieldChanges(changes...)
+	if err != nil {
+		err = fmt.Errorf("error marshaling changes for %s: %w", fd.Name(), err)
+		log.Error("failed to store field index changes", zap.Error(err))
+		return
+	}
+	if b == nil {
+		// No fields, nothing to do but close the file.
+		return
+	}
+	var fi os.FileInfo
+	if fi, err = fd.Stat(); err != nil {
+		err = fmt.Errorf("unable to get size of %s: %w", fd.Name(), err)
+		return
+	} else if fi.Size() > fscm.goodSize {
+		// If we had a partial write last time, truncate the file to remove it.
+		if err = fd.Truncate(fscm.goodSize); err != nil {
+			err = fmt.Errorf("cannot truncate %s to last known good size of %d after incomplete write: %w", fd.Name(), fscm.goodSize, err)
+			return
+		}
+	}
+
+	if _, err = writeSizePlusBuffer(fd, b); err != nil {
+		err = fmt.Errorf("failed writing to %s: %w", fd.Name(), err)
+		return
+	} else if fi, err = fd.Stat(); err != nil {
+		err = fmt.Errorf("unable to get final size of %s after appendation: %w", fd.Name(), err)
+		return
+	} else {
+		fscm.goodSize = fi.Size()
+	}
+}
+
+func writeSizePlusBuffer(w io.Writer, b []byte) (int, error) {
+	var numBuf [8]byte
+	binary.LittleEndian.PutUint64(numBuf[:], uint64(len(b)))
+	if n, err := w.Write(numBuf[:]); err != nil {
+		return n, err
+	} else {
+		m, err := w.Write(b)
+		return n + m, err
+	}
+}
+
+func readSizePlusBuffer(r io.Reader, b []byte) ([]byte, error) {
+	var numBuf [8]byte
+
+	if _, err := r.Read(numBuf[:]); err != nil {
+		return nil, err
+	}
+	size := int(binary.LittleEndian.Uint64(numBuf[:]))
+	if cap(b) < size {
+		b = make([]byte, size)
+	}
+	_, err := io.ReadAtLeast(r, b, size)
+	if err != nil {
+		return nil, err
 	}
-	return fs.marshalMeasurementFieldSetNoLock()
+	return b, nil
 }
 
 func (fs *MeasurementFieldSet) renameFile(path string) error {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
 
 	if err := file.RenameFile(path, fs.path); err != nil {
-		return err
+		return fmt.Errorf("cannot rename %s to %s: %w", path, fs.path, err)
 	}
 
-	if err := file.SyncDir(filepath.Dir(fs.path)); err != nil {
-		return err
+	dir := filepath.Dir(fs.path)
+	if err := file.SyncDir(dir); err != nil {
+		return fmt.Errorf("cannot synch direcotry %s: %w", dir, err)
 	}
 
 	return nil
 }
 
+// marshalMeasurementFieldSetNoLock: remove the fields.idx file if no fields
+// otherwise, copy the in-memory version into a protobuf to write to
+// disk
 func (fs *MeasurementFieldSet) marshalMeasurementFieldSetNoLock() (marshalled []byte, err error) {
+	if len(fs.fields) == 0 {
+		// If no fields left, remove the fields index file
```

Comment:
Why not just have a file containing a protobuf with an empty list? The current behaviour seems cleaner, unless there's a reason to change it I'm missing.

Reply:
The current behavior is to delete the file, it was just happening in an unobvious spot.